<h1 
  style="color:#207d06; border-bottom: 3px solid #207d06;"
</h1>
<h1 
  style="color:#165b33; border-bottom: 3px solid #165b33;"
</h1>
<h1 
  style="color:#146b3a; border-bottom: 3px solid #146b3a;"
</h1>
<h1 
  style="color:#f8b229; border-bottom: 3px solid #f8b229;"
</h1>
<h1 
  style="color:#ea4630; border-bottom: 3px solid #ea4630;"
</h1>
<h1 
  style="color:#bb2528; border-bottom: 3px solid #bb2528;"
</h1>
<h1 
  style="color:#4974a5; border-bottom: 3px solid #bb2528;"
>
  --- Day 4: Scratchcards --- Island Island
</h1>


<h2 
  style="color:#207d06; border-bottom: 3px solid #f8b229; border-top: 3px solid #ea4630;"
>
  Part One
</h2>


The gondola takes you up. Strangely, though, the ground doesn't seem to be coming with you; you're not climbing a mountain. As the circle of Snow Island recedes below you, an entire new landmass suddenly appears above you! The gondola carries you to the surface of the new island and lurches into the station.

As you exit the gondola, the first thing you notice is that the air here is much **warmer** than it was on Snow Island. It's also quite **humid**. Is this where the water source is?

The next thing you notice is an Elf sitting on the floor across the station in what seems to be a pile of colorful square cards.

"Oh! Hello!" The Elf excitedly runs over to you. "How may I be of service?" You ask about water sources.

"I'm not sure; I just operate the gondola lift. That does sound like something we'd have, though - this is **Island Island**, after all! I bet the **gardener** would know. He's on a different island, though - er, the small kind surrounded by water, not the floating kind. We really need to come up with a better naming scheme. Tell you what: if you can help me with something quick, I'll let you **borrow my boat** and you can go visit the gardener. I got all these [scratchcards](https://en.wikipedia.org/wiki/Scratchcard) as a gift, but I can't figure out what I've won."

The Elf leads you over to the pile of colorful cards. There, you discover dozens of scratchcards, all with their opaque covering already scratched off. Picking one up, it looks like each card has two lists of numbers separated by a vertical bar (`|`): a list of **winning numbers** and then a list of **numbers you have**. You organize the information into a table (your puzzle input).

As far as the Elf has been able to figure out, you have to figure out which of the **numbers you have** appear in the list of **winning numbers**. The first match makes the card worth **one point** and each match after the first **doubles** the point value of that card.

For example:
```
Card 1: 41 48 83 86 17 | 83 86  6 31 17  9 48 53
Card 2: 13 32 20 16 61 | 61 30 68 82 17 32 24 19
Card 3:  1 21 53 59 44 | 69 82 63 72 16 21 14  1
Card 4: 41 92 73 84 69 | 59 84 76 51 58  5 54 83
Card 5: 87 83 26 28 32 | 88 30 70 12 93 22 82 36
Card 6: 31 18 13 56 72 | 74 77 10 23 35 67 36 11
```

In the above example, card 1 has five winning numbers (`41`, `48`, `83`, `86`, and `17`) and eight numbers you have (`83`, `86`, `6`, `31`, `17`, `9`, `48`, and `53`). Of the numbers you have, four of them (`48`, `83`, `17`, and `86`) are winning numbers! That means `card 1` is worth `8` points (1 for the first match, then doubled three times for each of the three matches after the first).

 - Card 2 has two winning numbers (`32` and `61`), so it is worth `2 points`.
 - Card 3 has two winning numbers (`1` and `21`), so it is worth `2 points`.
 - Card 4 has one winning number (`84`), so it is worth `1 point`.
 - Card 5 has no winning numbers, so it is worth `no points`.
 - Card 6 has no winning numbers, so it is worth `no points`.

So, in this example, the Elf's pile of scratchcards is worth `13 points`.




<h5 
  style="color:#8fca6b; border-bottom: 1px solid #f8b229; border-top: 1px solid #f8b229;"
>
  Take a seat in the large pile of colorful cards. How many points are they worth in total?
</h5>

To begin, get your [puzzle input](https://adventofcode.com/2023/day/4/input).

Answer:
 

In [95]:
import pandas as pd

df = pd.read_csv('./data/Day4.csv', delimiter=r'[:|]', names=('card_id', 'card_nums', 'winning_nums'), engine='python')

def tokenize(text):
    # Define a function to split sentences into a list of words
    import re
    tokens = text.split()
    return tokens

# Tokenize the number strings into numbers
df['card_nums'] = df['card_nums'].apply(lambda x: set(tokenize(x)))
df['winning_nums'] = df['winning_nums'].apply(lambda x: set(tokenize(x)))
df.head()


card_id                                 card_nums  \
0  Card   1   {44, 65, 5, 29, 69, 10, 43, 27, 11, 24}   
1  Card   2  {45, 19, 34, 85, 31, 73, 43, 15, 36, 25}   
2  Card   3   {4, 59, 19, 46, 23, 42, 98, 57, 18, 75}   
3  Card   4  {44, 19, 56, 93, 32, 13, 92, 66, 27, 70}   
4  Card   5  {45, 61, 47, 13, 23, 80, 18, 91, 30, 25}   

                                        winning_nums  
0  {66, 43, 38, 14, 9, 10, 6, 89, 87, 22, 18, 34,...  
1  {32, 1, 43, 68, 74, 14, 92, 57, 36, 6, 31, 73,...  
2  {42, 21, 38, 98, 57, 3, 79, 59, 22, 18, 75, 39...  
3  {32, 1, 42, 66, 63, 38, 70, 53, 14, 13, 92, 27...  
4  {46, 29, 60, 58, 55, 68, 2, 61, 74, 13, 12, 88...

In [96]:
def find_winning_numbers(set_1, set_2):
  winning_nums = len(set_1.intersection(set_2))
  answer = 2**(winning_nums-1) if winning_nums > 0 else 0
  return answer

def find_matches(set_1, set_2):
  return len(set_1.intersection(set_2))


In [97]:
card_value = []
matches = []
for i in range(df.shape[0]):
  set_1 = df['card_nums'].iloc[i]
  set_2 = df['winning_nums'].iloc[i]

  card_value.append(find_winning_numbers(set_1, set_2))
  matches.append(find_matches(set_1, set_2))

df['matches'] = matches
df['card_value'] = card_value
print(f'The total winnings for scratchcards = {sum(card_value)}')
df.head()



The total winnings for scratchcards = 21919


card_id                                 card_nums  \
0  Card   1   {44, 65, 5, 29, 69, 10, 43, 27, 11, 24}   
1  Card   2  {45, 19, 34, 85, 31, 73, 43, 15, 36, 25}   
2  Card   3   {4, 59, 19, 46, 23, 42, 98, 57, 18, 75}   
3  Card   4  {44, 19, 56, 93, 32, 13, 92, 66, 27, 70}   
4  Card   5  {45, 61, 47, 13, 23, 80, 18, 91, 30, 25}   

                                        winning_nums  matches  card_value  
0  {66, 43, 38, 14, 9, 10, 6, 89, 87, 22, 18, 34,...        4           8  
1  {32, 1, 43, 68, 74, 14, 92, 57, 36, 6, 31, 73,...       10         512  
2  {42, 21, 38, 98, 57, 3, 79, 59, 22, 18, 75, 39...        8         128  
3  {32, 1, 42, 66, 63, 38, 70, 53, 14, 13, 92, 27...       10         512  
4  {46, 29, 60, 58, 55, 68, 2, 61, 74, 13, 12, 88...        8         128

Your puzzle answer was `21919`.


<h5 
  style="color:#f8b229;"
>
  The first half of this puzzle is complete! It provides one gold star: *
</h5>

<h2 
  style="color:#207d06; border-bottom: 3px solid #f8b229; border-top: 3px solid #ea4630;"
>
  Part Two
</h2>




Just as you're about to report your findings to the Elf, one of you realizes that the rules have actually been printed on the back of every card this whole time.

There's no such thing as "points". Instead, scratchcards only cause you to **win more scratchcards** equal to the number of winning numbers you have.

Specifically, you win **copies** of the scratchcards below the winning card equal to the number of matches. So, if card 10 were to have 5 matching numbers, you would win one copy each of cards 11, 12, 13, 14, and 15.

Copies of scratchcards are scored like normal scratchcards and have the same card number as the card they copied. So, if you win a copy of card 10 and it has 5 matching numbers, it would then win a copy of the same cards that the original card `10` won: cards `11`, `12`, `13`, `14`, and `15`. This process repeats until none of the copies cause you to win any more cards. (Cards will never make you copy a card past the end of the table.)

This time, the above example goes differently:
```
Card 1: 41 48 83 86 17 | 83 86  6 31 17  9 48 53
Card 2: 13 32 20 16 61 | 61 30 68 82 17 32 24 19
Card 3:  1 21 53 59 44 | 69 82 63 72 16 21 14  1
Card 4: 41 92 73 84 69 | 59 84 76 51 58  5 54 83
Card 5: 87 83 26 28 32 | 88 30 70 12 93 22 82 36
Card 6: 31 18 13 56 72 | 74 77 10 23 35 67 36 11
```

 - Card 1 has `four matching numbers`, so you win one copy each of the next four cards: cards `2`, `3`, `4`, and `5`.
 - *Your original* card 2 has `two` matching numbers, so you win one copy each of cards `3` and `4`.
 - *Your copy* of card 2 *also wins one copy each of cards `3` and `4`*.
 - Your four instances of card 3 (one original and three copies) have two matching numbers, so you win four copies each of cards 4 and 5.
 - Your eight instances of card 4 (one original and seven copies) have one matching number, so you win eight copies of card 5.
 - Your fourteen instances of card 5 (one original and thirteen copies) have no matching numbers and win no more cards.
 - Your one instance of card 6 (one original) has no matching numbers and wins no more cards.
Once all of the originals and copies have been processed, you end up with 1 instance of card 1, 2 instances of card 2, 4 instances of card 3, 8 instances of card 4, 14 instances of card 5, and 1 instance of card 6. In total, this example pile of scratchcards causes you to ultimately have 30 scratchcards!

Process all of the original and copied scratchcards until no more scratchcards are won. Including the original set of scratchcards, 

<h5 
  style="color:#8fca6b; border-bottom: 1px solid #f8b229; border-top: 1px solid #f8b229;"
>
  how many total scratchcards do you end up with?
</h5>


Answer:

<h5 style="text-align:right; padding-right: 10%;">
  <a href="#title">Top Of Page</a>
</h5>

In [105]:
len(df['matches'])


186

In [101]:
copies_dict[1] = 0
copies_dict


{1: 0}

In [106]:
copies_dict = {}
copies_dict[0] = 0

for row in range(186):
  match_count = df['matches'][row]
  b = copies_dict[row] + 1
  c = [row + z for z in range(y)]
  copies_dict[c[0]] + b

copies_dict


KeyError: 1